src/libraries/System.Linq/src/System/Linq/SegmentedArrayBuilder.cs

```diff
@@ -51,30 +52,44 @@ public void Dispose()
             int segmentsCount = _segmentsCount;
             if (segmentsCount != 0)
             {
-                ReadOnlySpan<T[]> segments = _segments;
+                ReturnArrays(segmentsCount);
+            }
+        }
 
-                // We need to return all rented arrays to the pool, and if the arrays contain any references,
-                // we want to clear them first so that the pool doesn't artificially root contained objects.
-                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
-                {
-                    // Return all but the last segment. All of these are full and need to be entirely cleared.
-                    foreach (T[] segment in segments.Slice(0, segmentsCount - 1))
-                    {
-                        ArrayPool<T>.Shared.Return(segment, clearArray: true);
-                    }
+        private void ReturnArrays(int segmentsCount)
+        {
+            Debug.Assert(segmentsCount > 0);
+            ReadOnlySpan<T[]> segments = _segments;
 
-                    // For the last segment, we can clear only what we know was used.
-                    T[] currentSegment = segments[segmentsCount - 1];
-                    Array.Clear(currentSegment, 0, _countInCurrentSegment);
-                    ArrayPool<T>.Shared.Return(currentSegment);
+            // We need to return all rented arrays to the pool, and if the arrays contain any references,
+            // we want to clear them first so that the pool doesn't artificially root contained objects.
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                // Return all but the last segment. All of these are full and need to be entirely cleared.
+                segmentsCount--;
+                foreach (T[] segment in segments.Slice(0, segmentsCount))
+                {
+                    Array.Clear(segment);
+                    ArrayPool<T>.Shared.Return(segment);
                 }
-                else
+
+                // For the last segment, we can clear only what we know was used.
+                T[] currentSegment = segments[segmentsCount];
+                Array.Clear(currentSegment, 0, _countInCurrentSegment);
+                ArrayPool<T>.Shared.Return(currentSegment);
+            }
+            else
+            {
+                // Return every rented array without clearing.
+                for (int i = 0; i < segments.Length; i++)
                 {
-                    // Return every rented array without clearing.
-                    foreach (T[] segment in segments.Slice(0, segmentsCount))
+                    T[] segment = segments[i];
+                    if (segment is null)
```

Comment:
What could cause the segment to be null here? Is the current annotation not correct?

Reply:
Every reference inside of a struct defaults to null.

Reply:
Should it be annotated to be `T[]?` instead?

Reply:
I don't think this one place matters. If we really wanted to, we could change the annotation inside of the InlineArray, and then every place that uses it. That's separate from this PR, which isn't changing anything to do with that.

This particular line is just using a null check on the array instead of slicing the span as a way to know when it's done iterating.